servant/src/Servant/Types/SourceT.hs

```diff
@@ -305,13 +306,28 @@ fromAction stop action = SourceT ($ fromActionStep stop action)
 {-# INLINE fromAction #-}
 
 fromActionStep :: Functor m => (a -> Bool) -> m a -> StepT m a
-fromActionStep stop action = loop where
-    loop = Effect $ fmap step action
-    step x
-        | stop x    = Stop
-        | otherwise = Yield x loop
+fromActionStep stop action = fromConsumableActionStep action' () where
+  action' = const $ flip fmap action $
+    \x -> if stop x
+          then Nothing
+          else Just (x,())
 {-# INLINE fromActionStep #-}
 
+-- | Create a `StepT' from a consumable @c@, that is to say from an input and an action that returns
+-- and is called again on an updated version of that input.
+fromConsumableActionStep :: Functor m
```

Comment:
I like the function, but the type immediately brings `unfold`-style functions in mind. This makes me think some name based around `unfold` might be clearer here? "Consumable" on the contrary doesn't really have a standard intuition we can hang onto.

Reply:
Thank you. Is `fromUnfoldActionStep` better?

Reply:
Yes that is better IMO. I'd perhaps even drop `from` and (`Action` or `Step`). `unfoldStep`? `unfoldAction`?